File: app/src/main/java/n7/ad2/heroes/HeroesFragment.java
package n7.ad2.heroes;


import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.paging.PagedList;

import android.content.Intent;
import androidx.databinding.DataBindingUtil;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.ActivityOptionsCompat;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.appcompat.widget.SearchView;

import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import n7.ad2.R;
import n7.ad2.databinding.FragmentHeroesBinding;
import n7.ad2.heroes.db.HeroModel;
import n7.ad2.heroes.full.HeroFullActivity;
import n7.ad2.main.MainViewModel;

import static n7.ad2.heroes.full.HeroFullActivity.HERO_CODE_NAME;
import static n7.ad2.heroes.full.HeroFullActivity.HERO_NAME;
import static n7.ad2.ui.MainActivity.LOG_ON_RECEIVE;


public class HeroesFragment extends Fragment implements SearchView.OnQueryTextListener {

    private HeroesViewModel viewModel;
    private FragmentHeroesBinding binding;
    private MainViewModel mainViewModel;

    public HeroesFragment() {
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_search, menu);
        super.onCreateOptionsMenu(menu, inflater);
        MenuItem searchHero = menu.findItem(R.id.action_search);
        SearchView searchView = (SearchView) searchHero.getActionView();
        searchView.setOnQueryTextListener(this);
    }

    public void startHeroFull(View view, HeroModel model) {
        Intent intent = new Intent(view.getContext(), HeroFullActivity.class);
        intent.putExtra(HERO_NAME, model.getName());
        intent.putExtra(HERO_CODE_NAME, model.getCodeName());
        if (getActivity() != null) {
            ActivityOptionsCompat options = ActivityOptionsCompat.makeSceneTransitionAnimation(getActivity(), view, "iv");
            startActivity(intent, options.toBundle());
        } else {
            startActivity(intent);
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_heroes, container, false);
        return binding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);


//        getActivity().setTitle(getString(R.string.ad2_toolbar, getString(R.string.heroes)));
        getActivity().setTitle(R.string.heroes);
        getActivity().sendBroadcast(new Intent(LOG_ON_RECEIVE).putExtra(LOG_ON_RECEIVE, "heroes_activity_created"));
        setRetainInstance(true);//фрагмент не уничтожается а передаётся новому активити (пропускает методы onCreate&onDestroy)
        setHasOptionsMenu(true);//вызов метода onCreateOptionsMenu в фрагменте

        mainViewModel = ViewModelProviders.of(getActivity()).get(MainViewModel.class);
        viewModel = ViewModelProviders.of(this).get(HeroesViewModel.class);

        initPagedListAdapter();
    }

    private void initPagedListAdapter() {
        binding.rvFragmentHeroes.setHasFixedSize(true);// если recyclerView не будет изменяться в размерах тогда ставим true
        binding.rvFragmentHeroes.setLayoutManager(new GridLayoutManager(getContext(), 3));
        binding.rvFragmentHeroes.setItemAnimator(new DefaultItemAnimator());
        final HeroesPagedListAdapter adapter = new HeroesPagedListAdapter(this); // PagedListAdapter, заточенный под чтение данных из PagedList.
        binding.rvFragmentHeroes.setAdapter(adapter);
        viewModel.getHeroesByFilter("").observe(this, new Observer<PagedList<HeroModel>>() {
            @Override
            public void onChanged(@Nullable PagedList<HeroModel> heroModels) {
                adapter.submitList(heroModels);
            }
        });

    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        return false;
    }

    @Override
    public boolean onQueryTextChange(String s) {
        final HeroesPagedListAdapter adapter = new HeroesPagedListAdapter(this);
        binding.rvFragmentHeroes.setAdapter(adapter);
        viewModel.getHeroesByFilter(s).observe(this, new Observer<PagedList<HeroModel>>() {
            @Override
            public void onChanged(@Nullable PagedList<HeroModel> heroModels) {
                adapter.submitList(heroModels);
            }
        });
        return true;
    }
}

File: app/src/main/java/n7/ad2/heroes/HeroesFragment.kt
package n7.ad2.heroes

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import androidx.appcompat.widget.SearchView
import androidx.core.app.ActivityOptionsCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.fragment.app.viewModels
import androidx.lifecycle.observe
import androidx.recyclerview.widget.GridLayoutManager
import n7.ad2.R
import n7.ad2.databinding.FragmentHeroesBinding
import n7.ad2.heroes.db.HeroModel
import n7.ad2.heroes.full.HeroFullActivity
import n7.ad2.main.MainViewModel
import n7.ad2.ui.MainActivity

class HeroesFragment : Fragment(R.layout.fragment_heroes) {

    private val viewModel: HeroesViewModel by viewModels()
    private lateinit var binding: FragmentHeroesBinding
    private val mainViewModel: MainViewModel by activityViewModels()

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        // implement search for last queries https://developer.android.com/guide/topics/search/adding-recent-query-suggestions
        inflater.inflate(R.menu.menu_search, menu)
        val searchHero = menu.findItem(R.id.action_search)
        val searchView = searchHero.actionView as SearchView

        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String?): Boolean {
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                viewModel.filterHeroes(newText)
                return true
            }
        })
    }

    fun startHeroFragment(view: View, model: HeroModel) {
        Intent(view.context, HeroFullActivity::class.java).apply {
            putExtra(HeroFullActivity.HERO_NAME, model.name)
            putExtra(HeroFullActivity.HERO_CODE_NAME, model.codeName)
            if (activity != null) {
                val options = ActivityOptionsCompat.makeSceneTransitionAnimation(requireActivity(), view, "iv")
                startActivity(this, options.toBundle())
            } else {
                startActivity(this)
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding = FragmentHeroesBinding.bind(view).also {
            it.lifecycleOwner = viewLifecycleOwner
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        requireActivity().apply {
            setTitle(R.string.heroes)
            sendBroadcast(Intent(MainActivity.LOG_ON_RECEIVE).putExtra(MainActivity.LOG_ON_RECEIVE, "heroes_activity_created"))
        }
        retainInstance = true //фрагмент не уничтожается а передаётся новому активити (пропускает методы onCreate&onDestroy)
        setHasOptionsMenu(true) //вызов метода onCreateOptionsMenu в фрагменте
        setupAdapter()
    }

    private fun setupAdapter() {
        val myAdapter = HeroesPagedListAdapter(this) // PagedListAdapter, заточенный под чтение данных из PagedList.
        binding.rv.apply {
            setHasFixedSize(true)
            layoutManager = GridLayoutManager(context, 3)
            adapter = myAdapter
        }
        viewModel.heroesPagedList.observe(viewLifecycleOwner, myAdapter::submitList)
    }
}